### PR TITLE
refactor: centralize tensor_s_growth handling in GrowingModule

### PR DESCRIPTION
**Pull request by stephane-rivaud**
- Move `tensor_s_growth` property and setter from `LinearGrowingModule` and `Conv2dGrowingModule` to parent class `GrowingModule`.
- Add proper `tensor_s_growth` initialization in FullConv2dGrowingModule since it is different from the `tensor_s` of the previous module.
- Remove explicit initialization of `tensor_s_growth` in `GrowingModule` as it is generally equal to the `tensor_s`of the previous module, the only exception being for `FullConv2dGrowingModule`.
- Remove redundant `s_growth_is_needed` flag, using isinstance checks instead
- Improve error handling for different types of previous modules

This refactoring improves code maintainability by reducing duplication and establishing a more consistent pattern for tensor growth handling across modules.

### CODE REVIEW

**Review by stelladk — APPROVED** [review]
Looking good!

File: src/gromo/modules/growing_module.py [hunk 1]
```diff
@@ -858,6 +846,37 @@ def tensor_s(self) -> TensorStatistic:
...
+            )
+
+    @tensor_s_growth.setter
+    def tensor_s_growth(self, value) -> None:
+        """
+        Allow to set the tensor_s_growth but has no effect.
+        """
+        raise AttributeError(
+            f"You tried to set tensor_s_growth of a GrowingModule (name={self.name})."
+            "This is not allowed as s growth is the same as tensor_s."
```
[the remark this answers is not in the document]
Comment (stelladk) [reply src/gromo/modules/growing_module.py:877]: I think we can ignore that at lest for now

[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/growing_module.py:877]: Agreed

File: src/gromo/modules/growing_module.py [hunk 2]
```diff
@@ -682,7 +669,8 @@ def forward(self, x):
         self.tensor_m.updated = False
         self.tensor_m_prev.updated = False
         self.cross_covariance.updated = False
-        if self.s_growth_is_needed:
+        if isinstance(self.previous_module, GrowingModule):
+            # TODO: change this condition by using self._allow_growing
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/growing_module.py:673]: This is ongoing development.

**Comment by stephane-rivaud** [comment]
## 🧪 Test Suite Refactoring Complete

### 📝 Overview
The test suite has been successfully refactored to align with the centralized `tensor_s_growth` handling implementation. All tests are passing and coverage is fully maintained.

### 🔄 Changes Summary

#### ✅ **Removed Outdated Tests** (`test_linear_growing_module.py`)
- Removed 4 obsolete `tensor_s_growth` tests that were no longer relevant:
  - `test_tensor_s_growth`
  - `test_tensor_s_growth_errors` 
  - `test_tensor_s_growth_property_coverage`
  - `test_tensor_s_growth_error_conditions`

#### ✅ **Added Centralized Tests** (`test_growing_module.py`)
- **Integrated 7 comprehensive `tensor_s_growth` tests** into the existing `TestGrowingModule` class:
  - `test_tensor_s_growth_no_previous_module`
  - `test_tensor_s_growth_with_growing_module_previous`
  - `test_tensor_s_growth_with_merge_growing_module_previous`
  - `test_tensor_s_growth_with_unsupported_previous_module`
  - `test_tensor_s_growth_setter_raises_error`
  - `test_tensor_s_growth_functional_behavior` (with bias parametrization)
  - `test_tensor_s_growth_multiple_layer_chain`

- **Removed separate `TestGrowingModuleTensorSGrowth` class** for better organization

#### ✅ **Added Specialized Tests** (`test_conv2d_growing_module.py`)
- **Added 4 `tensor_s_growth` tests** to `TestFullConv2dGrowingModule` class:
  - `test_tensor_s_growth_custom_implementation`
  - `test_tensor_s_growth_independence_from_previous_module`
  - `test_tensor_s_growth_shape_correctness`
  - `test_tensor_s_growth_update`

### 📊 Test Results & Coverage Analysis

| Metric | Result | Status |
|--------|--------|--------|
| **Total Tests Added** | 11 new `tensor_s_growth` tests | ✅ |
| **Total Tests Removed** | 4 obsolete tests | ✅ |
| **All Tests Passing** | 297/297 | ✅ |
| **Line Coverage** | 96.72% (2,328/2,407 lines) | ✅ |
| **Branch Coverage** | 87.24% (670/768 branches) | ✅ |
| **Combined Coverage** | 94.4% (2,998/3,175 total elements) | ✅ |
| **Code Quality** | All pre-commit hooks pass | ✅ |

#### 🔍 **Coverage Explanation**
The coverage metrics show two different measurements:
- **96.72% Line Coverage**: Percentage of source code lines executed by tests (standard metric)
- **94.4% Combined Coverage**: Weighted average of both line and branch coverage (terminal display)
- Both metrics are **identical to main branch** - no coverage regression

### 🎯 Key Improvements

1. **Better Test Organization**: `tensor_s_growth` tests are now logically grouped with their respective module classes instead of being isolated
2. **Comprehensive Coverage**: Tests cover all error conditions, functional behavior, and edge cases for the centralized implementation
3. **Code Quality**: Import statements moved to file tops following Python conventions
4. **Maintainability**: Reduced test class proliferation while maintaining full functionality

### 🔍 Verification

```bash
# GrowingModule tensor_s_growth tests
pytest tests/test_growing_module.py::TestGrowingModule -k "tensor_s_growth" -v
# 7 passed ✅

# FullConv2dGrowingModule tensor_s_growth tests  
pytest tests/test_conv2d_growing_module.py::TestFullConv2dGrowingModule -k "tensor_s_growth" -v
# 4 passed ✅

# Coverage verification
pytest --cov=src --cov-report=xml
# Line coverage: 96.72% (identical to main branch)
```

### 💡 Summary

The refactored test suite now properly reflects and validates the centralized `tensor_s_growth` design while maintaining identical coverage to the main branch. All tests pass and the code follows established patterns for better maintainability.

**Review by TheoRudkiewicz — APPROVED** [review]
Definitely a better solution than the flag ! Seems correctly changed in the three classes. Thank you !
